### PR TITLE
[Feature] GM State Change Persistance

### PR DESCRIPTION
**Pull request by fryguy503**
- Flymode and Invulnerable will now persist over zoning.
- Appended GMSpeed, Flymode and Invulnerable to the hideme message GMs see when they first login.
- Added #godmode [on/off] command to turn on or off hideme, flymode, gmspeed and invulnerable all in one shot.
- /becomenpc will now disable tells to the target player. It will also automatically disable GM States that interfere with its functionality.
- GM Command /toggle will not properly turn tells on/off
- GMs will now be notified if they are ignoring tells when they first zone-in, provided their GM flag is up.
- Added TellsOff variable to the output to #showstats

### CODE REVIEW

**Review by mackal — APPROVED** [review]

**Review by Kinglykrab — APPROVED** [review]
👍

File: world/zoneserver.cpp [hunk 1]
```diff
@@ -455,13 +455,10 @@ void ZoneServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p) {
 					!cle ||
 					cle->Online() < CLE_Status::Zoning ||
 					(
-						cle->TellsOff() &&
+						cle->TellsOff() &&
```
Comment (hgtw) [inline world/zoneserver.cpp:458]: trailing whitespace

File: world/zoneserver.cpp [hunk 2]
```diff
@@ -455,13 +455,10 @@ void ZoneServer::HandleMessage(uint16 opcode, const EQ::Net::Packet &p) {
...
 					(
-						cle->TellsOff() &&
+						cle->TellsOff() && 
 						(
-							(
-								cle->Anon() == 1 &&
-								scm->fromadmin < cle->Admin()
-							) ||
-							scm->fromadmin < AccountStatus::QuestTroupe
+							scm->fromadmin < cle->Admin()
```
Comment (hgtw) [inline world/zoneserver.cpp:460]: trailing whitespace

**Comment by Akkadius** [comment]
Made a handful of tweaks without creeping into refactoring a bunch of the account level stuff

**Comment by Akkadius** [comment]
Merge conflicts

File: utils/sql/db_update_manifest.txt [hunk 3]
```diff
@@ -444,6 +444,7 @@
 9188|2022_07_14_zone_expansion_revert.sql|SHOW COLUMNS FROM `zone` LIKE 'expansion'|empty|
 9189|2022_07_10_character_task_rewarded.sql|SHOW COLUMNS FROM `character_tasks` LIKE 'was_rewarded'|empty|
 9190|2022_07_13_task_reward_points.sql|SHOW COLUMNS FROM `tasks` LIKE 'reward_points'|empty|
+9191|2022_07_28_gm_state_changes.sql|SHOW COLUMNS FROM `account` LIKE `invulnerable`|empty|
```
Comment (hgtw) [inline utils/sql/db_update_manifest.txt:447]: Same typo I made in my last patch, need to surround the column name with single quotes not backticks

**Comment by Akkadius** [comment]
image

Changed string building routine to use a vector of strings

```cpp
	if (GetGM() && (GetHideMe() || GetGMSpeed() || GetGMInvul() || flymode != 0 || tellsoff)) {
		std::vector<std::string> state = {"currently"};
		if (GetHideMe()) {
			state.emplace_back("hidden to all clients");
		}
		if (GetGMSpeed()) {
			state.emplace_back("running at GM speed");
		}
		if (GetGMInvul()) {
			state.emplace_back("invulnerable to all damage");
		}
		if (flymode == 1) {
			state.emplace_back("flying");
		}
		else if (flymode == 2) {
			state.emplace_back("levitating");
		}
		if (tellsoff) {
			state.emplace_back("ignoring tells");
		}

		if (!state.empty()) {
			Message(Chat::Red, "[GM] You are %s.", Strings::Join(state, ", ").c_str());
		}
	}
```

**Review by hgtw — APPROVED** [review]

File: common/repositories/base/base_account_repository.h [hunk 4]
```diff
@@ -150,6 +159,9 @@ class BaseAccountRepository {
 		entry.ls_id          = "eqemu";
 		entry.lsaccount_id   = 0;
 		entry.gmspeed        = 0;
+		entry.invulnerable   = 0;
+		entry.flymode        = 0;
```
Comment (Kinglykrab) [inline common/repositories/base/base_account_repository.h:163]: Can use `GravityBehavior::Ground` here maybe?

File: zone/client_packet.cpp [hunk 5]
```diff
@@ -569,7 +570,31 @@ void Client::CompleteConnect()
...
+		if (GetHideMe()) {
+			state.emplace_back("hidden to all clients");
+		}
+		if (GetGMSpeed()) {
+			state.emplace_back("running at GM speed");
+		}
+		if (GetGMInvul()) {
+			state.emplace_back("invulnerable to all damage");
+		}
+		if (flymode == 1) {
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:584]: Can use `GravityBehavior::Flying` here maybe?

File: zone/client_packet.cpp [hunk 6]
```diff
@@ -569,7 +570,31 @@ void Client::CompleteConnect()
...
+			state.emplace_back("flying");
+		}
+		else if (flymode == 2) {
```
Comment (Kinglykrab) [inline zone/client_packet.cpp:587]: Can use `GravityBehavior::Levitating` here maybe?

File: common/shareddb.cpp [hunk 7]
```diff
@@ -65,40 +66,56 @@ SharedDatabase::~SharedDatabase() = default;
...
+
+bool SharedDatabase::SetGMFlymode(uint32 account_id, uint8 flymode)
+{
+	auto a = AccountRepository::FindOne(*this, account_id);
+	if (a.id > 0) {
+		a.flymode = flymode ? 1 : 0;
+		AccountRepository::UpdateOne(*this, a);
+	}
+
+	return a.id > 0;
```
Comment (Kinglykrab) [inline common/shareddb.cpp:118]: Does this allow flymodes other than just Fly Mode 1 (Flying)? There's 4 other types as well.

File: common/repositories/base/base_account_repository.h (hunk 4 above)
Comment (Akkadius) [reply common/repositories/base/base_account_repository.h:163]: Base repositories are code generated and are fairly dumb. This is to just represent "zero" or "default" values and wouldn't yield much improvement to readability here to try and inject constants

File: zone/client_packet.cpp (hunk 5 above)
Comment (Akkadius) [reply zone/client_packet.cpp:584]: Done

File: zone/client_packet.cpp (hunk 6 above)
Comment (Akkadius) [reply zone/client_packet.cpp:587]: Done

File: common/shareddb.cpp (hunk 7 above)
Comment (Akkadius) [reply common/shareddb.cpp:118]: Fixed this one

**Pull request merged**